api-publishing/src/main/java/org/zalando/nakadi/EventPublishingController.java

```diff
@@ -38,30 +40,105 @@
                                      final BlacklistService blacklistService,
                                      final NakadiKpiPublisher nakadiKpiPublisher) {
         this.publisher = publisher;
+        this.binaryPublisher = binaryPublisher;
         this.eventTypeMetricRegistry = eventTypeMetricRegistry;
         this.blacklistService = blacklistService;
         this.nakadiKpiPublisher = nakadiKpiPublisher;
     }
 
-    @RequestMapping(value = "/event-types/{eventTypeName}/events", method = POST)
-    public ResponseEntity postEvents(@PathVariable final String eventTypeName,
-                                     @RequestBody final String eventsAsString,
-                                     final HttpServletRequest request,
-                                     final Client client)
+    @RequestMapping(
+            value = "/event-types/{eventTypeName}/events",
+            method = POST,
+            consumes = "application/json"
+    )
+    public ResponseEntity postJsonEvents(@PathVariable final String eventTypeName,
+                                         @RequestBody final String eventsAsString,
+                                         final HttpServletRequest request,
+                                         final Client client)
             throws AccessDeniedException, BlockedException, ServiceTemporarilyUnavailableException,
             InternalNakadiException, EventTypeTimeoutException, NoSuchEventTypeException {
         return postEventsWithMetrics(eventTypeName, eventsAsString, request, client, false);
+    }
+
+    @RequestMapping(
+            value = "/event-types/{eventTypeName}/events",
+            method = POST,
+            consumes = "application/avro-binary; charset=utf-8",
+            produces = "application/json; charset=utf-8"
+    )
+    public ResponseEntity postBinaryEvents(@PathVariable final String eventTypeName,
+                                           @RequestBody final byte[] batch,
+                                           final HttpServletRequest request,
+                                           final Client client)
+            throws AccessDeniedException, BlockedException, ServiceTemporarilyUnavailableException,
+            InternalNakadiException, EventTypeTimeoutException, NoSuchEventTypeException {
+        try {
+            return postBinaryEvents(eventTypeName, batch, request, client, false);
+        } catch (IOException e) {
+            throw new InternalNakadiException("failed to parse batch", e);
+        }
+    }
+
+    private ResponseEntity postBinaryEvents(final String eventTypeName,
+                                            final byte[] batch,
+                                            final HttpServletRequest request,
+                                            final Client client,
+                                            final boolean delete) throws IOException {
+        TracingService.setOperationName("publish_events")
```

Comment:
Maybe use `publish_binary_event` so that we can see how each one perform using the tracing data.

Reply:
But this might actually break the checks we have now.
So maybe lets keep it as `publish_events`

Reply:
We can set "format" attribute on the span if we really want to tell apart binary from json.  Not sure we will ever be looking there, though.